### PR TITLE
Make WithContext conditional.

### LINKED ISSUE #202: Make WithContext() conditional.
Followup for https://github.com/census-instrumentation/opencensus-cpp/pull/200

### PR DESCRIPTION
**Pull request by g-easy**
While here, add an example in the comments.

### CODE REVIEW

File: opencensus/context/with_context.h [hunk 1]
```diff
@@ -21,15 +21,22 @@ namespace opencensus {
...
+// {
+//   WithContext wc(op.ctx_);
+//   // Do work.
+// }
 class WithContext {
  public:
-  explicit WithContext(const Context& ctx);
-  explicit WithContext(Context&& ctx);
+  explicit WithContext(const Context& ctx, bool cond = true);
+  explicit WithContext(Context&& ctx, bool cond = true);
```
Comment (bogdandrutu) [inline opencensus/context/with_context.h]: Does this need to be public?

File: opencensus/context/with_context.h [hunk 2]
```diff
@@ -39,6 +46,9 @@ class WithContext {
   WithContext& operator=(const WithContext&) = delete;
   WithContext& operator=(WithContext&&) = delete;
 
+  void ConditionalSwap();
+
+  const bool cond_;
```
Comment (bogdandrutu) [inline opencensus/context/with_context.h]: Order the members by size.

File: opencensus/context/with_context.h (hunk 1 above)
Comment (g-easy) [reply opencensus/context/with_context.h]: Do you mean WithContext in general? If we want users to be able to capture contexts and install them later without using Wrap() then yes.

File: opencensus/context/with_context.h (hunk 2 above)
Comment (g-easy) [reply opencensus/context/with_context.h]: Done.

File: opencensus/context/with_context.h (hunk 1 above)
Comment (g-easy) [reply opencensus/context/with_context.h]: Bogdan meant the rvalue reference constructor.

The reason for keeping this is: if the caller has saved a copy of a Context, and wants to install it, and won't need it later, they can std::move() it to save another copy.

**Pull request merged**